packages/protobuf-test/src/proto-delimited.test.ts

```diff
@@ -0,0 +1,216 @@
+        .finish();
+      it("should return size", function () {
+        const got = protoDelimited.peekSize(bytes);
+        expect(got.size).toBe(testMessagesSizes[0]);
+      });
+      it("should return expected offset", function () {
+        const got = protoDelimited.peekSize(bytes);
+        expect(got.offset).toBe(
+          bytes.byteLength - testMessagesBytes[0].byteLength
+        );
+      });
+    });
+    describe("used for parsing a stream", function () {
+      async function* createAsyncIterableBytes(
+        bytes: Uint8Array,
+        chunkSize = 2,
+        delay = 5
+      ): AsyncIterable<Uint8Array> {
+        let offset = 0;
+        for (;;) {
+          await new Promise((resolve) => setTimeout(resolve, delay));
+          const end = Math.min(offset + chunkSize, bytes.byteLength);
+          yield bytes.slice(offset, end);
+          offset = end;
+          if (offset === bytes.length) {
+            break;
+          }
+        }
+      }
+      const stream = createAsyncIterableBytes(
+        new BinaryWriter()
+          .uint32(testMessagesSizes[0])
+          .raw(testMessagesBytes[0])
+          .uint32(testMessagesSizes[1])
+          .raw(testMessagesBytes[1])
+          .finish()
+      );
+      /**
+       * An example implementation for a function decoding size-delimited messages
+       * from a simple stream using protoDelimited.peekSize and protoDelimited.dec.
+       * This example does not limit the amount of data read into memory.
+       */
+      async function* decStream<T extends Message<T>>(
+        iterable: AsyncIterable<Uint8Array>,
+        type: MessageType<T>
+      ) {
+        // append chunk to buffer, returning updated buffer
+        function append(buffer: Uint8Array, chunk: Uint8Array): Uint8Array {
+          const n = new Uint8Array(buffer.byteLength + chunk.byteLength);
+          n.set(buffer);
+          n.set(chunk, buffer.length);
+          return n;
+        }
+        let buffer = new Uint8Array(0);
+        for await (const chunk of iterable) {
+          buffer = append(buffer, chunk);
+          for (;;) {
+            const size = protoDelimited.peekSize(buffer);
+            if (size.eof) {
+              // size is incomplete, buffer more data
+              break;
+            }
+            if (size.offset + size.size > buffer.byteLength) {
+              // message is incomplete, buffer more data
+              break;
+            }
+            yield protoDelimited.dec(type, buffer);
+            buffer = buffer.subarray(size.offset + size.size);
+          }
+        }
+        if (buffer.byteLength > 0) {
+          throw new Error("incomplete data");
+        }
+      }
+      it("should decode messages", async function () {
+        const decoded: TestAllTypesProto3[] = [];
+        for await (const m of decStream(stream, TestAllTypesProto3)) {
+          decoded.push(m);
+        }
+        expect(decoded).toStrictEqual(testMessages);
+      });
```

Comment:
An example for using `protoDelimited.peekSize` to read from an `AsyncIterable<Uint8Array>`.

Reply:
This is surprisingly readable!